contracts/randomNumber/RandomNoManager.sol

```diff
@@ -0,0 +1,93 @@
+    IParameters public parameters;
+
+    event RandomNumberAvailable(uint32 epoch);
+
+    /// @param blockManagerAddress The address of the BlockManager Contract
+    /// @param parametersAddress The address of the Parameters contract
+    function initialize(address blockManagerAddress, address parametersAddress) external initializer onlyRole(DEFAULT_ADMIN_ROLE) {
+        grantRole(SECRETS_MODIFIER_ROLE, blockManagerAddress);
+        parameters = IParameters(parametersAddress);
+    }
+
+    /// @notice Allows Client to register for random number
+    /// Per request a rquest id is generated, which is binded to one epoch
+    /// this epoch is current epoch if Protocol is in commit state, and epoch + 1 if in any other states
+    /// @param requestId : unique request id
+    function register(bytes32 requestId) external override initialized {
+        uint32 epoch = getEpoch(parameters.epochLength());
+        State state = getState(parameters.epochLength());
+        require(requests[msg.sender][requestId] == 0, "Duplicate Request ID");
+        if (state == State.Commit) {
+            requests[msg.sender][requestId] = epoch;
+        } else {
+            requests[msg.sender][requestId] = epoch + 1;
+        }
+    }
+
+    /// @notice Called by BlockManager in ClaimBlockReward or ConfirmBlockLastEpoch in confirm state
+    /// @param epoch current epoch
+    /// @param _secret hash of encoded rando secret from stakers
+    function provideSecret(uint32 epoch, bytes32 _secret) external override onlyRole(SECRETS_MODIFIER_ROLE) {
+        /// @dev this require is added for extra assurance to clients,
+        /// to give them assurance that once secret is set for epoch, it cant be changed
+        /// as admin could always override this SECRETS_MODIFIER_ROLE role
+        require(secrets[epoch] == 0x0, "Secret already set");
+        secrets[epoch] = _secret;
+        emit RandomNumberAvailable(epoch);
+    }
+
+    /// @notice Allows Client to check first if given req id is avaialble before registering.
+    /// @param requestId : unique request id
+    function isReqIdAvailable(bytes32 requestId) external view override initialized returns (bool) {
+        return (requests[msg.sender][requestId] == 0);
+    }
+
+    /// @notice Allows client to pull random number once available
+    /// Random no is generated from secret of that epoch and request id, its unique per requestid
+    /// @param requestId : A unique id per request
+    /// @return random number
+    function getRandomNumber(bytes32 requestId) external view override returns (uint256) {
+        uint32 epochOfRequest = requests[msg.sender][requestId];
+        return _generateRandomNumber(epochOfRequest, requestId);
+    }
+
+    /// @notice Allows client to get generic random number of last epoch
+    /// @return random number
+    function getGenericRandomNumberOfLastEpoch() external view override returns (uint256) {
```

Comment:
Probably can be called getGenericRandomNumber and the last epoch part can be mentioned in documentation.

Reply:
Just saw `getGenericRandomNumber` Why two methods?

Reply:
one allows to get genericRandomNumber of any epoch, if client wants to , there epoch needs to be passed
one allows to get genericRandomNumber of last epoch, their client doesn't have to pass anything

Reply:
I meant, `getGenericRandomNumberOfLastEpoch` and `getGenericRandomNumber` are pretty much the same, can we have just one `getGenericRandomNumber`? Are these different use cases for getting a random number for this epoch and the previous epoch?

Reply:
hmm yes 

```getGenericRandomNumber```
whenever client starts a game, they can store what current epoch was in their contract 
and at time of pulling lever can call ```getGenericRandomNumber(epoch)```

```getGenericRandomNumberLastEpoch```
This allows use case where random number is irrespective of epoch, just fetch latest random number 
We can say that client can fetch the current epoch on itself and call ```getGenericRandomNumber(epoch-1)```
That way also execution cost remains same, there is no extra execution cost on our side, bothways client needs to do work.
Via separate method, its just better UX

Reply:
Ok, sounds good.